src/attribution/onINP.ts

```diff
@@ -25,38 +31,228 @@ import {
+  PerformanceEventTiming,
+  DOMHighResTimeStamp
+> = new WeakMap();
+
+// A reference to the idle task used to clean up entries from the above
+// variables. If the value is -1 it means no task is queue, and if it's
+// greater than -1 the value corresponds to the idle callback handle.
+let idleHandle: number = -1;
+
+/**
+ * Adds new LoAF entries to the `pendingLoAFs` list.
+ */
+const handleLoAFEntries = (entries: PerformanceLongAnimationFrameTiming[]) => {
+  entries.forEach((entry) => pendingLoAFs.push(entry));
+};
+
+/**
+ * Groups entries that were presented within the same animation frame by
+ * a common `renderTime`. This function works by referencing
+ * `pendingEntriesGroupMap` and using an existing render time if one is found
+ * (otherwise creating a new one). This function also adds all interaction
+ * entries to an `entryToRenderTimeMap` WeakMap so that the "grouped" render
+ * times can be looked up later.
+ */
+const groupEntriesByRenderTime = (entries: PerformanceEventTiming[]) => {
+  entries.forEach((entry) => {
+    let renderTime = entry.startTime + entry.duration;
+    let previousRenderTime;
+
+    // Iterate of all previous render times in reverse order to find a match.
+    // Go in reverse since the most likely match will be at the end.
+    for (let i = previousRenderTimes.length - 1; i >= 0; i--) {
+      previousRenderTime = previousRenderTimes[i];
+
+      // If a previous render time is within 8ms of the current render time,
+      // assume they were part of the same frame and re-use the previous time.
+      // Also break out of the loop because all subsequent times will be newer.
+      if (Math.abs(renderTime - previousRenderTime) <= 8) {
+        const group = pendingEntriesGroupMap.get(previousRenderTime)!;
+        group.startTime = Math.min(entry.startTime, group.startTime);
+        group.processingStart = Math.min(
+          entry.processingStart,
+          group.processingStart,
+        );
+        group.processingEnd = Math.max(
+          entry.processingEnd,
+          group.processingEnd,
+        );
+        group.entries.push(entry);
+
+        renderTime = previousRenderTime;
+        break;
+      }
+    }
+
+    // If there was no matching render time, assume this is a new frame.
+    if (renderTime !== previousRenderTime) {
+      previousRenderTimes.push(renderTime);
+      pendingEntriesGroupMap.set(renderTime, {
+        startTime: entry.startTime,
+        processingStart: entry.processingStart,
+        processingEnd: entry.processingEnd,
+        entries: [entry],
+      });
+    }
+
+    // Store the grouped render time for this entry for reference later.
+    if (entry.interactionId || entry.entryType === 'first-input') {
+      entryToRenderTimeMap.set(entry, renderTime);
+    }
+  });
+
+  // Queue cleanup of entries that are not part of any INP candidates.
+  if (idleHandle < 0) {
+    idleHandle = whenIdle(cleanupEntries);
+  }
+};
+
+const cleanupEntries = () => {
```

Comment:
@mmocny if you have time to review this function, that would be helpful. This function isn't part of the public API, but if any part of this logic fails (or is incorrect) then it could result in issues, so I'd love a second pair of eyes on it.

In short, this function tries to remove references to `event` and LoAF entries that are no longer needed (so we're not just storing everything in memory), but there's the potential that I could remove a reference to something that will later been needed and then it won't get reported in the attribution object (or worse, there will be an error because other code will try to reference a property of an object that no longer exists).